erts/doc/src/persistent_term.xml

```diff
@@ -0,0 +1,278 @@
+    deleted, the process will be garbage collected and the term copied
+    to process.</p>
+
+    <p>Avoid updating or deleting more than one persistent term at a
+    time.  Each deleted term will trigger its own global GC. That
+    means that deleting N terms will make the system less responsive N
+    times longer than deleting a single persistent term. Therefore,
+    terms that are to be updated at the same time should be collected
+    into a larger term, for example, a map or a tuple.</p>
+  </section>
+
+  <section>
+    <title>Example</title>
+
+    <p>The following example shows how lock contention for ETS tables
+    can be minimized by having one ETS table for each scheduler. The
+    table identifiers for the ETS tables are stored as a single
+    persistent term:</p>
+
+<pre>
+    %% There is one ETS table for each scheduler.
+    Sid = erlang:system_info(scheduler_id),
+    Tid = element(Sid, persistent_term:get(?MODULE)),
+    ets:update_counter(Tid, Key, 1).</pre>
+
+  </section>
+
+  <datatypes>
+    <datatype>
+      <name name="key"/>
+      <desc>
+        <p>Any Erlang term.</p>
+      </desc>
+    </datatype>
+    <datatype>
+      <name name="value"/>
+      <desc>
+        <p>Any Erlang term.</p>
+      </desc>
+    </datatype>
+  </datatypes>
+
+  <funcs>
+    <func>
+      <name name="erase" arity="1"/>
+      <fsummary>Erase the name for a persistent term.</fsummary>
+      <desc>
+        <p>Erase the name for the persistent term with key
+	<c><anno>Key</anno></c>. The return value will be <c>true</c>
+	if there was a persistent term with the key
+	<c><anno>Key</anno></c>, and <c>false</c> if there was no
+	persistent term associated with the key.</p>
+	<p>If there existed a previous persistent term associated with
+	key <c><anno>Key</anno></c>, a global GC has been initiated
+	when <c>erase/1</c> returns. See <seealso
+	marker="#description">Description</seealso>.</p>
+      </desc>
+    </func>
+
+    <func>
+      <name name="get" arity="0"/>
+      <fsummary>Get all persistent terms.</fsummary>
+      <desc>
+        <p>Retrieve the keys and values for all persistent terms.
+	The keys will be copied to the heap for the process calling
+	<c>get/0</c>, but the values will not.</p>
+      </desc>
+    </func>
+
+    <func>
+      <name name="get" arity="1"/>
+      <fsummary>Get the value for a persistent term.</fsummary>
+      <desc>
+        <p>Retrieve the value for the persistent term associated with
+        the key <c><anno>Key</anno></c>. The lookup will be made in
+	constant time and the value will not be copied to the heap
+	of the calling process.</p>
+	<p>This function fails with a <c>badarg</c> exception if no
+	term has been stored with the key
+	<c><anno>Key</anno></c>.</p>
+	<p>If the calling process holds on to the value of the
+	persistent term and the persistent term is deleted in the future,
+	the term will be copied to the process.</p>
```

Comment:
So if you have a max heap size set to the process, this can end up killing the process for memory allocations even if it is not running, right?

I currently don't see a test for that one, but it sounds like the kind of thing that could really go wrong if not done well -- although I'm not familiar with the max-heap-size implementation for checks, it just doesn't necessarily sound like something I'd imagine triggered during a GC in normal circumstances.

Reply:
Good catch! I didn't think of the max heap size checks.

The max heap size checks *are* triggered during a GC (it would be too expensive to do the checks at other times).

It seems that the max heap size limit checks are not triggered by the particular kind of GC that copies the term into the process. Not sure whether this intentional. Since this kind of GC is used also when purging old code, I am not sure that we should that we should start triggering the max heap size check in OTP 21.2. I'll discuss this with the OTP team.

Reply:
I have opened #1991 to make sure the max heap size limit checks triggers by the global GC that copies literals into processes. We have reached the conclusion that it would be safe to include that bug fix in OTP 21.2.